channeldb/kvdb/etcd/stm.go

```diff
@@ -303,24 +295,14 @@ func (rs readSet) gets() []v3.Op {
 	return ops
 }
 
-// cmps returns a cmp list testing values in read set didn't change.
-func (rs readSet) cmps(lset map[string]interface{}) []v3.Cmp {
-	if len(lset) > 0 {
-		cmps := make([]v3.Cmp, 0, len(lset))
-		for key := range lset {
-			if getValue, ok := rs[key]; ok {
-				cmps = append(
-					cmps,
-					v3.Compare(v3.ModRevision(key), "=", getValue.rev),
-				)
-			}
-		}
-		return cmps
-	}
-
+// cmps returns a compare list which will serve as a precondition testing that
+// the values in the read set didn't change.
+func (rs readSet) cmps() []v3.Cmp {
 	cmps := make([]v3.Cmp, 0, len(rs))
 	for key, getValue := range rs {
-		cmps = append(cmps, v3.Compare(v3.ModRevision(key), "=", getValue.rev))
+		cmps = append(cmps, v3.Compare(
+			v3.ModRevision(key), "=", getValue.rev,
```

Comment:
i notice that the etcd stm objects use a slightly different comparison, in that they also check for a 0 value when no values are returned https://github.com/etcd-io/etcd/blob/master/clientv3/concurrency/stm.go#L360

is there a reason we don't do the same? or perhaps it is done implicitly in how `.rev` is set?

Reply:
Yeah, we do the same jut differently. This is why we insert the key with zero rev and no value (which is then included in the compare set). Also moved that from `Get` to the `fetch` to catch range start/ends too!